File: src/main/java/domain/model/kfzw/Kfzw.java
package domain.model.kfzw;

import domain.math.Index;
import domain.math.LinearExtrapolation;

import java.util.Arrays;

public class Kfzw {

    public static Double[][] generateKfzw(Double[] xAxisOld, Double[][] kfzwOld, Double[] xAxisNew) {

        Double[][] kwzwNew = new Double[kfzwOld.length][kfzwOld[0].length];

        for (int i = 0; i < kfzwOld.length; i++) {
            for (int j = 0; j < kfzwOld[i].length; j++) {

                int indexKey = Index.getInsertIndex(Arrays.asList(xAxisOld), xAxisNew[j]);

                double x0;
                double x1;
                double y0;
                double y1;

                if (indexKey < kfzwOld[i].length - 1) {
                    x0 = xAxisOld[indexKey];
                    x1 = xAxisOld[indexKey + 1];
                    y0 = kfzwOld[i][indexKey];
                    y1 = kfzwOld[i][indexKey + 1];
                } else {
                    x0 = xAxisOld[indexKey - 1];
                    x1 = xAxisOld[indexKey];
                    y0 = kfzwOld[i][indexKey - 1];
                    y1 = kfzwOld[i][indexKey];
                }

                double value = Math.max(LinearExtrapolation.extrapolate(new Double[]{x0, x1}, new Double[]{y0, y1}, xAxisNew[j]), -13.5);

                kwzwNew[i][j] = value;
            }
        }

        return kwzwNew;
    }
}

File: src/main/kotlin/domain/model/kfzw/Kfzw.kt
package domain.model.kfzw

import domain.math.Index
import domain.math.LinearExtrapolation

object Kfzw {
    fun generateKfzw(xAxisOld: Array<Double>, kfzwOld: Array<Array<Double>>, xAxisNew: Array<Double>): Array<Array<Double>> {
        return Array(kfzwOld.size) { i ->
            Array(kfzwOld[0].size) { j ->
                var indexKey = Index.getInsertIndex(xAxisOld.toList(), xAxisNew[j])

                if (indexKey > 0 && indexKey < kfzwOld[i].size - 1 && xAxisNew[j] < xAxisOld[indexKey]) {
                    indexKey--
                }

                val x0: Double
                val x1: Double
                val y0: Double
                val y1: Double

                if (indexKey < kfzwOld[i].size - 1) {
                    x0 = xAxisOld[indexKey]
                    x1 = xAxisOld[indexKey + 1]
                    y0 = kfzwOld[i][indexKey]
                    y1 = kfzwOld[i][indexKey + 1]
                } else {
                    x0 = xAxisOld[indexKey - 1]
                    x1 = xAxisOld[indexKey]
                    y0 = kfzwOld[i][indexKey - 1]
                    y1 = kfzwOld[i][indexKey]
                }

                maxOf(LinearExtrapolation.extrapolate(arrayOf(x0, x1), arrayOf(y0, y1), xAxisNew[j]), -13.5)
            }
        }
    }
}
